cpp/src/hash/sha_hash.cu

```diff
@@ -64,7 +64,7 @@ CUDA_DEVICE_CALLABLE uint64_t swap_endian(uint64_t x)
 
 }  // namespace
 
-template <typename IntermediateType, typename WordType, uint32_t MessageChunkSize>
+template <typename Hasher, typename IntermediateType, typename WordType, uint32_t MessageChunkSize>
```

Comment:
My gut reaction is that the additional template parameters should be instead derived/queried from the `Hasher` type as that feels like a more descriptive encapsulation of the parameters of a `Hasher`.

Reply:
Unless of course the parameters like `WordType` and `MessageChunkSize` can vary independently from `Hasher`.

Reply:
I re-implemented this template to only depend on `Hasher`, which encapsulates the other template parameters.